docs/src/references/synchronize-daemons.md

```diff
@@ -0,0 +1,116 @@
+ software list on `tedge/commands/req/software/list`.
+- Now the tedge-agent processes this request and sends the software list to the mapper, once the mapper gets the response it will forward it to the c8y cloud.
+
+Example 2,
+- When tedge-mapper-c8y is started, it will subscribe to `tedge/health/#` topic
+- When c8y-log-plugin is started, it will publish health status as `up` on to `tedge/health/c8y-log-plugin` topic
+- Now the mapper picks up this message and creates the c8y_LogfileRequest operation file in `/etc/tedge/operations/c8y`.
+- Also the mapper creates the `c8y-log-plugin.toml` file, which contains the `supported log types` in `/etc/tedge/c8y`.
+- Mapper sends the updated supported operations list to the c8y cloud. Also, it sends the supported `log types`(118,software-management) to the c8y cloud.
+
+Pros: 
+-   This will remove the file system dependency when the daemons want to create the supported operation files on the mapper file system.
+    For example, tedge-agent wants to create the supported operations files in /etc/tedge/operations/c8y directory.
+
+Cons: 
+-   This leads to hard dependency on the list of daemons in the mapper. For example, the tedge-mapper-c8y has to subscribe to tedge/health/tedge-agent
+topic to receive the health status from the tedge-agent. Based on this the mapper has to create a supported operation file in /etc/tedge/operations/c8y.
+-   This will be a hard-coded dependency of the name of the plugin and can’t be updated on the fly if someone wants to change the name of the daemon
+    or wants to add a new daemon.
+
+## Proposal 2:  Introduce an `init message` mechanism
+
+Make all the tedge components publish a `retained init` message describing their role (software management, config management, etc)
+ and other details specific to that role.
+
+The proposed topic will be `tedge/<device-id>/init/<cloud>`.
+Here `<cloud>` could be c8y, azure, aws, etc
+
+The message template is {Operation_name: <content>, Operation_name:<content>, Config_types: type1..typen}.
+Here `Operation_name` is the name of the operation to be published to the cloud and 
+the `content` is the content of the operation file, if any.
+Here `Config_types` is the configuration types that are supported by that particular daemon.
+
+Example 1:
+- When the tedge-mapper-c8y is started, it will subscribe to `tedge/<device-id>/init/#` topic
+- When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`
+- Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdat` supported operations file in `/etc/tedge/operations/c8y`. 
+- Mapper sends the updated supported operations list to the c8y cloud.
+- Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the
+ software list on `tedge/commands/req/software/list`.
```

Comment:
> When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`

If this message is empty, what the difference with a health status?

> Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdat` supported operations file in `/etc/tedge/operations/c8y`.

This is okay the mapper has to know the mapping from tedge capabilities to c8y capability.

> Mapper sends the updated supported operations list to the c8y cloud.

We need to be sure that the operations published on the cloud are not  added/removed each time the agent is started/stopped. I guess the intent is to have these capabilities published/removed on install/uninstall on this new topic.

> Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the
 software list on `tedge/commands/req/software/list`.

This is where the proposed approach to capability discovery falls short. The mapper *still* has to know that the `c8y_SoftwareUpdate` capability is provided via  the `tedge/commands/req/software/list` topic. Somehow the mapper gains no more information with a init capability message than with a simple name as 'tegde-agent' is installed.

Being aware that a capability is available is not enough. The mapper needs to know how to use it, to invoke the capability and to translate the messages. A well-known name is enough to trigger the associated code.

Reply:
> > When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`
> 
> If this message is empty, what the difference with a health status?

Yes, this is only for the `tedge-agent` case. Because it will not send any capabilities.
But this can be changed later and ask the `tedge-agent` to publish the capabilities.

> 
> > Now the mapper picks up this message and creates the `c8y_Restart, c8y_SoftwareUpdat` supported operations file in `/etc/tedge/operations/c8y`.
> 
> This is okay the mapper has to know the mapping from tedge capabilities to c8y capability.
> 
> > Mapper sends the updated supported operations list to the c8y cloud.
> 
> We need to be sure that the operations published on the cloud are not added/removed each time the agent is started/stopped. I guess the intent is to have these capabilities published/removed on install/uninstall on this new topic.

If we just stick to publishing the capabilities on `install/uninstall` then the dependency is MQTT broker must be running during installation.

So, I feel we have to make a trade-off here, even if the agent sends the capabilities when it starts. The c8y mapper will try creating and publishing the same set of capabilities.

> 
> > Also, now that the tedge-agent is up and running, the tedge-mapper-c8y will publish the request to get the
> > software list on `tedge/commands/req/software/list`.
> 
> This is where the proposed approach to capability discovery falls short. The mapper _still_ has to know that the `c8y_SoftwareUpdate` capability is provided via the `tedge/commands/req/software/list` topic. Somehow the mapper gains no more information with a init capability message than with a simple name as 'tegde-agent' is installed.

Agree, the plan here is to still keep the mapper cloud-specific. So, it will continue to do the same thing as it is doing as of now except that creating capabilities on the right directories and publishing those capabilities to the cloud.
The main advantage here is the `plugin` need not worry about creating the `operations` file. This will be helpful when each plugin is containerized and run in different containers.
> 
> Being aware that a capability is available is not enough. The mapper needs to know how to use it, to invoke the capability, and to translate the messages. A well-known name is enough to trigger the associated code.

Reply:
>> > When tedge-agent is started, it will publish `{}` an empty `init` message on to `tedge/init/c8y/tedge-agent`
>> 
>> If this message is empty, what the difference with a health status?
>
> Yes, this is only for the `tedge-agent` case. Because it will not send any capabilities.
> But this can be changed later and ask the `tedge-agent` to publish the capabilities.

The fact that things can work with an empty payload is an argument for my point: a capability is captured with a well-known name that *implies* a set of operations, message schema and API calls.

In this example, the mapper has to know that `tedge-agent` is a daemon that can do two things: restart the device and install software. By the way, it would be good to distinguish the daemon from the capabilities. Here, it would be better to have two: `device-restart` and `software-update`.